optaplanner-examples/src/main/java/org/optaplanner/examples/common/experimental/impl/ConsecutiveIntervalInfoImpl.java

```diff
@@ -0,0 +1,214 @@
+    private final BiFunction<PointType_, PointType_, DifferenceType_> differenceFunction;
+    private final Iterable<IntervalBreak<IntervalType_, PointType_, DifferenceType_>> breaksIterable;
+
+    public ConsecutiveIntervalInfoImpl(TreeSet<IntervalSplitPoint<IntervalType_, PointType_>> splitPointSet,
+            BiFunction<PointType_, PointType_, DifferenceType_> differenceFunction) {
+        clusterStartSplitPointToCluster = new TreeMap<>();
+        clusterStartSplitPointToNextBreak = new TreeMap<>();
+        intervalClusterIterable = new MapValuesIterable<>(clusterStartSplitPointToCluster);
+        breaksIterable = new MapValuesIterable<>(clusterStartSplitPointToNextBreak);
+        this.splitPointSet = splitPointSet;
+        this.differenceFunction = differenceFunction;
+    }
+
+    protected void addInterval(Interval<IntervalType_, PointType_> interval) {
+        NavigableMap<IntervalSplitPoint<IntervalType_, PointType_>, IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_>> intersectedIntervalClusterMap =
+                clusterStartSplitPointToCluster.subMap(
+                        ObjectUtils.defaultIfNull(clusterStartSplitPointToCluster.floorKey(interval.getStartSplitPoint()),
+                                interval.getStartSplitPoint()),
+                        true, interval.getEndSplitPoint(), true);
+
+        // Case: the interval cluster before this interval does not intersect this interval
+        if (!intersectedIntervalClusterMap.isEmpty()
+                && intersectedIntervalClusterMap.get(intersectedIntervalClusterMap.firstKey()).getEndSplitPoint()
+                        .compareTo(interval.getStartSplitPoint()) < 0) {
+            intersectedIntervalClusterMap = intersectedIntervalClusterMap.subMap(intersectedIntervalClusterMap.firstKey(),
+                    false, intersectedIntervalClusterMap.lastKey(), true);
+        }
+
+        if (intersectedIntervalClusterMap.isEmpty()) {
+            IntervalSplitPoint<IntervalType_, PointType_> start = splitPointSet.floor(interval.getStartSplitPoint());
+            IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_> newCluster =
+                    new IntervalClusterImpl<>(splitPointSet, differenceFunction, start);
+            clusterStartSplitPointToCluster.put(start, newCluster);
+            Map.Entry<IntervalSplitPoint<IntervalType_, PointType_>, IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_>> nextClusterEntry =
+                    clusterStartSplitPointToCluster.higherEntry(start);
+            if (nextClusterEntry != null) {
+                clusterStartSplitPointToNextBreak.put(start,
+                        new IntervalBreakImpl<IntervalType_, PointType_, DifferenceType_>(newCluster,
+                                nextClusterEntry.getValue(), differenceFunction.apply(
+                                        newCluster.getEnd(), nextClusterEntry.getValue().getStart())));
```

Comment:
This is another piece of code that would benefit from declaring a bit more variables.
Build the statement incrementally, to make it easier to read.



    ... nextClusterEntryValue = nextClusterEntry.getValue();
    DIfferenceType_ difference = differenceFunction.apply(newCluster.getEnd(), nextClusterEntryValue.getStart());
    IntervalBreakImpl<...> newBreak = new IntervalBreakImpl<IntervalType_, PointType_, DifferenceType_>(newCluster, nextClusterEntryValue, difference);
    clusterStartSplitPointToNextBreak.put(start, newBreak);

See how much easier it is to read?